### PR TITLE
Add more description for from_raw_parts's unsafety

### LINKED ISSUE #26737: {String,Vec}::from_raw_parts docs need to be much clearer about incorrectness
The docs for these are currently fairly meagre:

>     unsafe fn from_raw_parts(buf: *mut u8, length: usize, capacity: usize) -> String

> Creates a new String from a length, capacity, and pointer.

> This is unsafe because:

> -    We call `Vec::from_raw_parts` to get a `Vec<u8>`;
> -    We assume that the Vec contains valid UTF-8.

>     unsafe fn from_raw_parts(ptr: *mut T, length: usize, capacity: usize) -> Vec<T>

> Creates a `Vec<T>` directly from the raw components of another vector.

> This is highly unsafe, due to the number of invariants that aren't checked.

They need to be more specific about the details, e.g.: 

- `ptr` needs to have been previously allocated via `String`/`Vec` (at least, it's highly likely to be incorrect if it wasn't)
- `capacity` needs to be the capacity that the pointer was allocated with.

Violating these may cause problems like corrupting the allocators internal datastructures.

### PR DESCRIPTION
**Pull request by steveklabnik**
Fixes #26737.

### CODE REVIEW

File: src/libcollections/string.rs [hunk 1]
```diff
@@ -317,9 +317,14 @@ impl String {
...
     ///
-    /// This is unsafe because:
+    /// # Unsafety
     ///
-    /// * We call `Vec::from_raw_parts` to get a `Vec<u8>`;
+    /// This is _very_ unsafe because:
+    ///
+    /// * We call `Vec::from_raw_parts` to get a `Vec<u8>`. Therefore, this
+    ///   function inherits all of its unsafety, see [its
+    ///   documentation](../collections/string/struct.String.html#method.from_raw_parts)
```
Comment (huonw) [inline src/libcollections/string.rs]: Should this link to `std`?

Comment (steveklabnik) [inline src/libcollections/string.rs]: Ahh, possibly, yes. I just looked at the URLs and compared them... also this is to `String`, not `Vec`, which I looked at like five times, sigh

**Pull request merged**